### PR TITLE
[#24789][prism] Handlers for combine, ParDo, GBK, Flatten 

### PR DESCRIPTION
**Pull request by lostluck**
Add in the handlers for Combine, ParDo, GBK, Flatten
Adds in an executor interface. 

No unit tests largely because these are largely rote Proto transformations, and are best tested in the pipeline context. They'll fully covered in the completed internal package.

See https://github.com/apache/beam/issues/24789 for more information.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @johannaojeling 

This is where things get interesting! This converts/prepares the Pipeline Proto representations to fully expanded alternatives and similar.  You'll probably find the comments on how SDFs and Combine composites are expanded interesting.

GroupByKeys and Flattens are pretty straight foward so far and do the appropriate in memory work. That would change depending on how data services evolve (eg. Stop being strictly in memory). But as it ism this is an OK abstraction to start.

The code living in an internal package gives us the freedom to evolve the abstractions we're using as needed.

**Comment by brucearctor** [comment]
LGTM - but I've only been skimming the code here [ and https://github.com/apache/beam/pull/25556 and https://github.com/apache/beam/pull/25557 ... and your other related/previous PRs ].  Fantastic contribution, so glad you've gotten it this far already!

**Review by johannaojeling — COMMENTED** [review]
Agreed, this is all excellent!

File: sdks/go/pkg/beam/runners/prism/internal/handlepardo.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	if err := (proto.UnmarshalOptions{}).Unmarshal(pardoPayload, pdo); err != nil {
+		panic(fmt.Sprintf("unable to decode ParDoPayload for transform[%v]", t.GetUniqueName()))
+	}
+
+	// Lets check for and remove anything that makes things less simple.
+	if pdo.OnWindowExpirationTimerFamilySpec == "" &&
+		!pdo.RequestsFinalization &&
+		!pdo.RequiresStableInput &&
+		!pdo.RequiresTimeSortedInput &&
+		//	len(pdo.SideInputs) == 0 &&
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/runners/prism/internal/handlepardo.go]: No. I have a comment elsewhere about side inputs.

File: sdks/go/pkg/beam/runners/prism/internal/handlerunner.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+// This file retains the logic for the pardo handler
+
+// RunnerCharacteristic holds the configuration for Runner based transforms,
+// such as GBKs, Flattens.
+type RunnerCharacteristic struct {
+	SDKFlatten bool // Sets whether we should force an SDK side flatten.
+	SDKGBK     bool // Sets whether the GBK should be handled by the SDK, if possible by the SDK.
+}
+
+func Runner(config any) *runner {
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/runners/prism/internal/handlerunner.go:48]: In a public API, returning unexported types is annoying for users, but it's less so under the isolation of the internal folder. It's likely the handlers will end up in their own package, and get exported out that way, with the types changed as appropriate.

While I've got the interfaces (preparer, executer) they're there to reduce the chance the abstraction will be broken. They'll be refined as other features are added to Prism, like Fusion, and handling the TestStream, and PubSubm URNs (and more).

**Pull request merged**